src/FwdState.cc

```diff
@@ -1290,7 +1288,7 @@ FwdState::logReplyStatus(int tries, const Http::StatusCode status)
 bool
 FwdState::exhaustedTries() const
 {
-    return n_tries >= Config.forward_max_tries;
+    return tries() >= Config.forward_max_tries;
```

Comment:
This change affects reforwarding logic AFAICT. Consider a situation where the destination has two IP addresses, IPv4 and IPv6. Let's assume forward_max_tries is set to 2. The first address is slow to respond, so Happy Eyeballs makes two TCP connection establishment attempts. Let's say TCP/IPv4 connection wins after all, and Squid sends a request to the IPv4 address. If the server response needs to be reforwarded, FwdState will check exhaustedTries() and

* Before this change: FwdState reforwards the request to the TCP/IPv6 connection, right?
* After this change: FwdState does not reforward the request.

Are there any flaws in the above analysis? If there are not, what do you propose as a fix?

Reply:
> Before this change: FwdState reforwards the request to the TCP/IPv6 connection, right?

No, I think that it won't reforward. If HappyConnOpener makes two connection attempts (as in your example), HappyConnOpener::n_tries becomes 2 and then passed over to FwdState::n_tries. If, after that, reforward is required, FwdState::exhaustedTries() returns false and no reforward is performed.
Basically, the changes introduced in this PR only substitute one counter with another (i.e, HappyConnOpener::n_tries with AccessLogEntry::requestAttempts), while leaving the forwarding behavior unchanged.

Reply:
Aha, so we are essentially dealing with buggy official code. The changes in this PR do not introduce the bug we discovered. The question is: Are they making the existing bug _worse_ by making the future fix harder?

I think there are at least answers to that question:

1. "yes" or "maybe": These changes essentially equate connecting attempts with forwarding attempts when, in fact, the two classes of attempts are distinct. The distinction is in increment _timing_ -- the second counter value is bounded by the first one, but the second counter is incremented _later_, and that delay is meaningful rather than accidental. FwdState reforwarding checks should use the second, delayed counter. We should not merge the two counters together.
2. "no, but this PR should remove the broken FwdState (and tunnel.cc) checks": One ALE-based counter is fine if FwdState relies on HappyConnOpener to observe the limit. If HappyConnOpener gave FwdState a connection, then the limit has already been observed (if not, then this is a HappyConnOpener bug unrelated to this PR).

The second answer would make everything simpler. Please check whether it is possible to fully outsource forward_max_tries checks to HappyConnOpener? In other words, are there some fundamental reasons why FwdState (and tunnel.cc) must maintain their own attempts counter (or equivalent)? For example, are there any connections that FwdState does not receive from HappyConnOpener _and_ that are subject to the forward_max_tries limit?

Reply:
> are there any connections that FwdState does not receive from HappyConnOpener 

'pinned' connections are separate and FwdState::usePinned() increments the counter.

Reply:
Yes, but, IIRC, there can be at most one PINNED connection, and PINNED destination(s) are tried first, before everything else. If that recollection is correct, and assuming forward_max_tries cannot be legitimately used to deny the very first forwarding attempt, then PINNED connections do not need forward_max_tries checks.

Reply:
> PINNED destination(s) are tried first, before everything else

It looks like that is true. However, failed pinned connections may be retried (FwdState::pinnedCanRetry()) and we need to check for forward_max_tries before picking another destination. We could do this probably in HappyConnOpener before attempting to connect (e.g., somewhere in HappyConnOpener::maybeOpenAnotherPrimeConnection()) but this looks awkward - in this case, we do not need to create HappyConnOpener at all.

Reply:
I am not sure which solution will be less awkward, but I think the discussion about the location of the missing check is a bit premature/misplaced. AFAICT, we agree that HappyConnOpener (and/or surrounding code) has a bug -- it may reach the forward_max_tries limit but then hide the last (spare) connection from the requestor, effectively depriving the requestor from the last connection attempt.

If we agree on that, then I think that bug should be fixed in a separate PR. This PR should assume that the bug is fixed. We can either pause this PR (until that other bug is fixed) or, with that assumption in mind, adjust (remove the duplicated check from FwdState/tunnel.cc) and merge this PR. If the latter does not make things worse, then I would prefer it because it allows us to make progress _and_ may make fixing the bug easier. What do you think?

Reply:
There is another counter for the same purpose: PeerSelector::foundPaths. It is incremented in PeerSelector::handlePath() in both pinned and regular cases and checked against forward_max_tries in PeerSelector::wantsMoreDestinations(). It looks like this check should be enough to cover forward_max_tries condition since it is performed earlier than similar/equivalent checks in FwdState and HappyConnOpener. In other words, if foundPaths reached its maximum, there are no new destinations and HappyConnOpener should not bother about it. If so, we can remove forward_max_tries checks from all places, except PeerSelector. Meanwhile, in this case, we can go on with this PR: AccessLogEntry::requestAttempts is merely for attempts counting, without any validation purpose.

Reply:
PeerSelector::foundPaths is not sufficient for forward_max_tries enforcement because FwdState (and, in the future, tunnel.cc) may retry the same found path twice. Look for the retryPath() call. You are right that PeerSelector::foundPaths should not exceed forward_max_tries, but, due to pconn races, FwdState may reach the attempts limit even before trying all the found paths.

Reply:
I am still not sure whether a single ALE-based counter is what we need long-term. I would not be surprised if fixing the lost destination bug discovered during this PR [review](https://github.com/squid-cache/squid/pull/541#discussion_r366872547) requires splitting that counter somehow. I recommend pausing this PR until we know what that fix looks like. I am marking this PR as "WIP" for now.

Reply:
>  I recommend pausing this PR until we know what that fix looks like.

The proposed fix is now posted as #1093. This PR should still wait (now for that PR landing).

Reply:
_Status update_: PR 1093 was merged. We need to resume this PR work. @eduard-bagdasaryan, if you think that we should resolve current merge conflicts before finalizing the fix design or if they are easy to resolve, please resolve them.

Reply:
I am resolving this old thread. The "making future bug fix hard" concern I raised here was addressed by PR 1093 (which fixed the bug this discussion exposed).